File: zipkin-server/src/test/java/zipkin2/server/internal/throttle/ThrottledStorageComponentTest.java
/*
 * Copyright The OpenZipkin Authors
 * SPDX-License-Identifier: Apache-2.0
 */
package zipkin2.server.internal.throttle;

import brave.Tracing;
import com.linecorp.armeria.common.metric.NoopMeterRegistry;
import org.junit.jupiter.api.Test;
import zipkin2.Component;
import zipkin2.internal.Nullable;
import zipkin2.server.internal.throttle.ThrottledStorageComponent.ThrottledSpanConsumer;
import zipkin2.storage.InMemoryStorage;
import zipkin2.storage.StorageComponent;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.AssertionsForClassTypes.assertThatExceptionOfType;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

class ThrottledStorageComponentTest {
  InMemoryStorage delegate = InMemoryStorage.newBuilder().build();
  @Nullable Tracing tracing;
  NoopMeterRegistry registry = NoopMeterRegistry.get();

  @Test void spanConsumer_isProxied() {
    ThrottledStorageComponent throttle =
      new ThrottledStorageComponent(delegate, registry, tracing, 1, 2, 1);

    assertThat(ThrottledSpanConsumer.class)
      .isSameAs(throttle.spanConsumer().getClass());
  }

  @Test void createComponent_withZeroSizedQueue() {
    int queueSize = 0;
    new ThrottledStorageComponent(delegate, registry, tracing, 1, 2, queueSize);
    // no exception == pass
  }

  @Test void createComponent_withNegativeQueue() {
    assertThatExceptionOfType(IllegalArgumentException.class).isThrownBy(() -> {
      int queueSize = -1;
      new ThrottledStorageComponent(delegate, registry, tracing, 1, 2, queueSize);
    });
  }

  /**
   * The {@code toString()} of {@link Component} implementations appear in health check endpoints.
   * Since these are likely to be exposed in logs and other monitoring tools, care should be taken
   * to ensure {@code toString()} output is a reasonable length and does not contain sensitive
   * information.
   */
  @Test void toStringContainsOnlySummaryInformation() {
    assertThat(new ThrottledStorageComponent(delegate, registry, tracing, 1, 2, 1))
      .hasToString("Throttled{InMemoryStorage{}}");
  }

  @Test void delegatesCheck() {
    StorageComponent mock = mock(StorageComponent.class);

    new ThrottledStorageComponent(mock, registry, tracing, 1, 2, 1).check();
    verify(mock, times(1)).check();
  }
}

File: zipkin-server/src/test/kotlin/zipkin2/server/internal/throttle/ThrottledStorageComponentTest.kt
/*
 * Copyright 2015-2019 The OpenZipkin Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package zipkin2.server.internal.throttle

import com.linecorp.armeria.common.metric.NoopMeterRegistry
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import zipkin2.storage.InMemoryStorage
import zipkin2.storage.StorageComponent

class ThrottledStorageComponentTest {
  val delegate = InMemoryStorage.newBuilder().build()
  val registry = NoopMeterRegistry.get()

  @Test fun spanConsumer_isProxied() {
    val throttle = ThrottledStorageComponent(delegate, registry, 1, 2, 1)

    assertThat(throttle.spanConsumer().accept(listOf()))
      .isInstanceOf(ThrottledCall::class.java)
  }

  @Test fun createComponent_withZeroSizedQueue() {
    val queueSize = 0
    ThrottledStorageComponent(delegate, registry, 1, 2, queueSize)
    // no exception == pass
  }

  @Test(expected = IllegalArgumentException::class)
  fun createComponent_withNegativeQueue() {
    val queueSize = -1
    ThrottledStorageComponent(delegate, registry, 1, 2, queueSize)
  }

  @Test fun niceToString() {
    assertThat(ThrottledStorageComponent(delegate, registry, 1, 2, 1))
      .hasToString("Throttled(InMemoryStorage{traceCount=0})")
  }

  @Test fun delegatesCheck() {
    val mock = mock(StorageComponent::class.java)
    ThrottledStorageComponent(mock, registry, 1, 2, 1).check()
    verify(mock, times(1)).check()
  }
}
